### PR TITLE
[CI] Build pytorch wheel with Torch XPU Operators on Windows

### PR DESCRIPTION
**Pull request by XinyueLunaFan**
# Description
This pipeline enables the CI build on Windows with PR labeled with ciflow/xpu. This will build torch binary with Torch XPU Operators on Windows using Vision Studio BuildTools 2022.

# Changes
1. Install xpu batch file (install_xpu.bat) - Check if build machine has oneAPI in environment, and if the version of it is latest. If not, install the latest public released oneAPI in the machine.
2. GHA callable pipeline (_win-build.yml) - Set vc_year and use_xpu as parameter to set build wheel environment.
3.  GHA workflow (xpu.yml) - Add a new windows build job and pass parameters to it.
4.  Build wheels script (.ci/pytorch/win-test-helpers/build_pytorch.bat) - Prepare environment for building, e.g. install oneAPI bundle.

# Note
1. For building wheels on Intel GPU, you need Vision Studio BuildTools version >= 2022
2. This pipeline requires to use Vision Studio BuildTools 2022 to build wheels. For now, we specify "windows.4xlarge.nonephemeral" as build machine label in the yaml file. We will request to add self-hosted runners with Intel GPU and Vision Studio BuildTools 2022 installed soon.

Work for #114850

### CODE REVIEW

File: .github/workflows/windows-build-xpu.yml [hunk 1]
```diff
@@ -0,0 +1,156 @@
...
+           rmdir /s /q "pytorch"
+         )
+         git clone https://github.com/pytorch/pytorch.git pytorch > git_pytorch_repo_log.txt
+         cd pytorch
+         call conda install conda-forge::rust -y
+         call conda install typing_extensions -y
+         python -m pip install -r requirements.txt > pip_install_pytorch_requirements_log.txt
+         python setup.py bdist_wheel > build_torch_wheel_log.txt
+         
+     - name: Build TorchVision
```
Comment (min-jean-cho) [inline .github/workflows/windows-build-xpu.yml:85]: Why TorchVision build is part of `ciflow/xpu`?

Comment (XinyueLunaFan) [reply .github/workflows/windows-build-xpu.yml]: Good point, I have removed build torchvision job from this script and will add it into UT script.

File: .ci/pytorch/win-test-helpers/installation-helpers/install_oneapi_bundle.bat [hunk 2]
```diff
@@ -0,0 +1,37 @@
+set ONEAPI_PARENT_DIR=C:\Program Files (x86)\Intel
```
Comment (chuanqi129) [inline .ci/pytorch/win-test-helpers/installation-helpers/install_oneapi_bundle.bat:1]: Do we need install xpu UMD driver on the Windows build runner?

File: .github/workflows/windows-build-xpu-nightly.yml [hunk 3]
```diff
@@ -0,0 +1,102 @@
+name: windows-build-xpu-nightly
```
Comment (chuanqi129) [inline .github/workflows/windows-build-xpu-nightly.yml:1]: Let's delete this workflow and focus on ci firstly, the CD has a series dedicate scripts and workflows

File: .github/workflows/windows-build-xpu.yml [hunk 4]
```diff
@@ -0,0 +1,128 @@
+name: windows-build-xpu
+
+on:
+  push:
+    tags:
+      - ciflow/xpu/*
```
Comment (chuanqi129) [inline .github/workflows/windows-build-xpu.yml:6]: Please move the xpu windows build as a job in https://github.com/pytorch/pytorch/blob/main/.github/workflows/xpu.yml

File: .github/workflows/windows-build-xpu.yml [hunk 5]
```diff
@@ -0,0 +1,128 @@
...
+env:
+    conda_env: 'xpu_venv'
+    conda_path: 'C:\Jenkins\Miniconda3'
+    python_version: '3.10'
+    USE_XPU: 1
+    VS2022INSTALLDIR: 'C:\Program Files (x86)\Microsoft Visual Studio\2022\BuildTools'
+
+jobs:
+  Preparation:
+    runs-on: ["${{ inputs.win-runner }}"]
```
Comment (chuanqi129) [inline .github/workflows/windows-build-xpu.yml:36]: Do we want to reuse existed windows runner? If so, could we reuse the workflow
https://github.com/pytorch/pytorch/blob/main/.github/workflows/_win-build.yml
and scripts
https://github.com/pytorch/pytorch/blob/main/.ci/pytorch/win-build.sh
directly with limited modification to add xpu support?

**Comment by chuanqi129** [comment]
> [CLA Missing ID](https://api.easycla.lfx.linuxfoundation.org/v2/repository-provider/github/sign/29311126/65600975/133151/#/?version=2) [CLA Not Signed](https://api.easycla.lfx.linuxfoundation.org/v2/repository-provider/github/sign/29311126/65600975/133151/#/?version=2)
> 
> * ✅ login: XinyueLunaFan  ([f768010](https://github.com/pytorch/pytorch/commit/f7680104091f294ff5370cbc0ee69d75df40f6ee), [62e3925](https://github.com/pytorch/pytorch/commit/62e392535a00566909702f53fa038c15d5432d69), [c26128e](https://github.com/pytorch/pytorch/commit/c26128ec7fa46d17b7043a093f7357fa5f599a1e), [fdcc9f1](https://github.com/pytorch/pytorch/commit/fdcc9f1fafb795b1cf1ec275b9615175174ee012), [93d6fbe](https://github.com/pytorch/pytorch/commit/93d6fbe47894f922054d80e91eb53a306a170133), [0ed6b74](https://github.com/pytorch/pytorch/commit/0ed6b7405793bfd7db52adecf87fb81fd1d0b9e0), [f6c3a3e](https://github.com/pytorch/pytorch/commit/f6c3a3e109f8422d8491eda0f6651bfe178882f7), [4491341](https://github.com/pytorch/pytorch/commit/44913416d3a3426acf3372c7a8b00d804999ebc9), [aedbc43](https://github.com/pytorch/pytorch/commit/aedbc43163de45aa63617b231cb29ecc234e61f2))
> * ❌ The email address for the commit ([d4a409f](https://github.com/pytorch/pytorch/commit/d4a409fbe625cda77b47bb817b1e4f79b73ad70c), [28a8d41](https://github.com/pytorch/pytorch/commit/28a8d410ef98221c992782751c5a83165d6ad8d2), [f2e5f4e](https://github.com/pytorch/pytorch/commit/f2e5f4e3982c323d3238aa6b21e5272356fbed9b), [9cb80d6](https://github.com/pytorch/pytorch/commit/9cb80d6bd82f6e841320fe2ce7eb8b04f622f5d4), [d27e83e](https://github.com/pytorch/pytorch/commit/d27e83e571b56abf5a6845c88040dc50ac905d9c), [20c9956](https://github.com/pytorch/pytorch/commit/20c9956bb33131d25d10cd3f5b13ea1084d1b27b), [b33ad22](https://github.com/pytorch/pytorch/commit/b33ad22e0bb7f9671505866ca9e23068e3614473), [8f1243e](https://github.com/pytorch/pytorch/commit/8f1243e90f270261765636e7d8b07c95e578518b), [d458d1b](https://github.com/pytorch/pytorch/commit/d458d1ba892ff9645d6ea21897fc938172d39b53), [89a229a](https://github.com/pytorch/pytorch/commit/89a229ae58f587c94b6658f7a1633353c33766ad), [0f3c0e0](https://github.com/pytorch/pytorch/commit/0f3c0e0ab20d9db4d179929cba6c7d73c8806e88), [7f85f69](https://github.com/pytorch/pytorch/commit/7f85f6944f4897831751c4f5a06cd0130dd46bb8), [ae2581e](https://github.com/pytorch/pytorch/commit/ae2581e931380c88c35f6cfafcaec79f7ec20915), [0b9889f](https://github.com/pytorch/pytorch/commit/0b9889f1d8912a33647385d99b6a4229a68ed6e8), [c30b3cf](https://github.com/pytorch/pytorch/commit/c30b3cfd0904234510afb80446cf697a919b4ce7), [5e81f83](https://github.com/pytorch/pytorch/commit/5e81f831ca570a720f450447ed66446df6b95db1), [0bd7010](https://github.com/pytorch/pytorch/commit/0bd7010ad0828285e47f5085f2be618f282906c6), [cc0242b](https://github.com/pytorch/pytorch/commit/cc0242b282a76017c95114dfe03506753e597b6a), [a4ab957](https://github.com/pytorch/pytorch/commit/a4ab95796be785883063f657ccf9f59ba8142d0d), [c7a3761](https://github.com/pytorch/pytorch/commit/c7a37616223968f939cfb9c977ce1096a797b6b3), [99c1725](https://github.com/pytorch/pytorch/commit/99c172538cef64bfd7d22ea4028e7bff885e287c), [20a7dbb](https://github.com/pytorch/pytorch/commit/20a7dbb48aaa95e4cd5edefc207ffe2a2f63d441), [e5597ff](https://github.com/pytorch/pytorch/commit/e5597ffbeea18fde06d235d7bb15d7abb62c7182), [8a50f8e](https://github.com/pytorch/pytorch/commit/8a50f8e8e47e145deab98ac4ecbaceac70a3aae6), [667d38f](https://github.com/pytorch/pytorch/commit/667d38fccf417825605581067dfaf046476ab680), [913399b](https://github.com/pytorch/pytorch/commit/913399b417d3ff03bc783c4f1e254eaf84417ce9), [95a80b6](https://github.com/pytorch/pytorch/commit/95a80b6dbeffd277d6b00dc3f5d3f4fd704b8dc2), [8099a97](https://github.com/pytorch/pytorch/commit/8099a97884774bd1ffadeff1aae51c329ec69688), [1a17c3d](https://github.com/pytorch/pytorch/commit/1a17c3d46ade214a85aa6c979826afd3f0b11d71), [7f8ed34](https://github.com/pytorch/pytorch/commit/7f8ed3422ee912697821de6b784140728c42093a), [302c920](https://github.com/pytorch/pytorch/commit/302c920b7eec56ebceecda6fc2892f216504016d), [8751b0d](https://github.com/pytorch/pytorch/commit/8751b0d89043f097c6f91b0eaed1c39f97a597d3), [b4a0337](https://github.com/pytorch/pytorch/commit/b4a0337e595cc8e82bc8d22ef786a54b9ecb0d2a), [b038600](https://github.com/pytorch/pytorch/commit/b0386005c1c5720b00f8103b1901da9ccf639e02), [d299122](https://github.com/pytorch/pytorch/commit/d299122d7e011b4c7c4958d295306bc4d8169f08), [c576c23](https://github.com/pytorch/pytorch/commit/c576c233a0f62f8876e25df2227fc1925fe28260), [fe79227](https://github.com/pytorch/pytorch/commit/fe7922791a1272b133b24e152f010a5f77781751), [87c4404](https://github.com/pytorch/pytorch/commit/87c44044a8763feecfac1e6cfb34a843e753d51d), [60b20a5](https://github.com/pytorch/pytorch/commit/60b20a515a872506493a052426a6cdc1e6e2df93), [3441603](https://github.com/pytorch/pytorch/commit/3441603b5f073ee682909d56ea1816d053cc8dbb), [8c21d3c](https://github.com/pytorch/pytorch/commit/8c21d3ca04c89463c0a925ef746a6b706e5f6be6), [a45560b](https://github.com/pytorch/pytorch/commit/a45560b44d22eec063a82af1be9ae3c644fa0f0b), [81952d1](https://github.com/pytorch/pytorch/commit/81952d11a01a92ca4c273b130d876887394dc183), [ef8efb7](https://github.com/pytorch/pytorch/commit/ef8efb760de23bff79f75bc4e70f7908fa9c468e), [facaf34](https://github.com/pytorch/pytorch/commit/facaf34317962dd31b6ee328c2b72dc50c8ed6cb), [085bd86](https://github.com/pytorch/pytorch/commit/085bd86ad4cdaad37ab6ddbd70e88a140341f63a), [b9e8f60](https://github.com/pytorch/pytorch/commit/b9e8f607f632013184a62a284f94679bb579a62c), [620cbb2](https://github.com/pytorch/pytorch/commit/620cbb285a8bd63da59de4c60767a0226a4220cd)) is not linked to the GitHub account, preventing the EasyCLA check. Consult [this Help Article](https://confluence.linuxfoundation.org/pages/viewpage.action?pageId=86641302) and [GitHub Help](https://help.github.com/en/github/committing-changes-to-your-project/why-are-my-commits-linked-to-the-wrong-user) to resolve. (To view the commit's email address, add .patch at the end of this PR page's URL.) For further assistance with EasyCLA, [please submit a support request ticket](https://jira.linuxfoundation.org/servicedesk/customer/portal/4).

@XinyueLunaFan please squash the commits as one with right user/email information

File: .ci/pytorch/win-test-helpers/installation-helpers/install_oneapi_bundle.bat [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+set INTEL_ONEAPI_PYTORCH_BUNDLE_URL=https://registrationcenter-download.intel.com/akdlm/IRC_NAS/5ca2021d-dd1a-4ab1-bd52-758fe63cf827/w_intel-for-pytorch-gpu-dev_p_0.5.2.19_offline.exe
```
Comment (min-jean-cho) [inline .ci/pytorch/win-test-helpers/installation-helpers/install_oneapi_bundle.bat:2]: What's the timeline to uplift intel-for-pytorch-gpu-dev to 0.5.3?

File: .github/workflows/windows-build-xpu.yml [hunk 7]
```diff
@@ -0,0 +1,128 @@
...
+          vision\build_vision_wheel_log.txt
+
+    - name: Store PyTorch Build Artifacts on S3
+      uses: seemethere/upload-artifact-s3@v5
+      with:
+        name: ${{ inputs.build-environment }}
+        retention-days: 14
+        if-no-files-found: error
+        path: |
+          pytorch\dist\torch-*.whl
```
Comment (min-jean-cho) [inline .github/workflows/windows-build-xpu.yml:119]: Do existing runners upload binaries?

File: .github/workflows/windows-build-xpu.yml [hunk 8]
```diff
@@ -0,0 +1,128 @@
...
+      uses: actions/upload-artifact@v2
+      with:
+        name: logs
+        path: |
+          pytorch\git_pytorch_repo_log.txt
+          pytorch\pip_install_pytorch_requirements_log.txt
+          pytorch\build_torch_wheel_log.txt
+          vision\git_vision_repo_log.txt
+          vision\pip_install_vision_requirements_log.txt
+          vision\build_vision_wheel_log.txt
```
Comment (min-jean-cho) [inline .github/workflows/windows-build-xpu.yml:110]: Ditto

File: .github/workflows/windows-build-xpu.yml [hunk 9]
```diff
@@ -0,0 +1,128 @@
...
+    runs-on: ["${{ inputs.win-runner }}"]
+    needs: [Preparation,Build-whls]
+    steps:
+    - name: Upload pytorch wheel
+      uses: actions/upload-artifact@v2
+      with:
+        name: pytorch wheel
+        path: 'pytorch\dist'
+
+    - name: Upload torchvision wheel
```
Comment (min-jean-cho) [inline .github/workflows/windows-build-xpu.yml:94]: Why TorchVision is needed for XPU CI flow?

File: .github/workflows/windows-build-xpu.yml [hunk 10]
```diff
@@ -0,0 +1,128 @@
...
+  Build-whls:
+    runs-on: ["${{ inputs.win-runner }}"]
+    needs: [Preparation]
+    steps:
+     - name: Build Torch
+       shell: cmd
+       run: |
+         call %conda_path%\Scripts\activate.bat %conda_env%
+         call "C:\Program Files (x86)\Intel\oneAPI\setvars.bat"
+         if exist "pytorch" (
```
Comment (min-jean-cho) [inline .github/workflows/windows-build-xpu.yml:74]: Is this needed?

**Comment by min-jean-cho** [comment]
Do we need to build against all default AOT targets for CI/CD?

File: .github/workflows/windows-build-xpu.yml (hunk 4 above)
Comment (chuanqi129) [reply .github/workflows/windows-build-xpu.yml:6]: Let's check whether we can reuse existed callable workflow _win-build.yml and build scripts directly. If can't, we also need to modify the new added windows build workflow as a callable workflow

File: .github/workflows/windows-build-xpu.yml (hunk 5 above)
Comment (chuanqi129) [reply .github/workflows/windows-build-xpu.yml:36]: Let's check the root cause of the MSVC 2019 build failure. And we also are able to check the existed windows runner whether has msvc 2022 installed or if it's possible to install to 2019 and 2022 in advance for those existed runner

File: .github/workflows/windows-build-xpu.yml (hunk 9 above)
Comment (XinyueLunaFan) [reply .github/workflows/windows-build-xpu.yml]: Deleted. Thanks for pointing out

File: .github/workflows/windows-build-xpu.yml (hunk 10 above)
Comment (XinyueLunaFan) [reply .github/workflows/windows-build-xpu.yml:74]: This is to get latest pytorch master and clear last build files.

File: .github/workflows/windows-build-xpu.yml (hunk 7 above)
Comment (XinyueLunaFan) [reply .github/workflows/windows-build-xpu.yml:110]: Yes, they put all things in a zip file and upload the zip file. I am working on to see if I can directly use their existing build configurations.

File: .github/workflows/windows-build-xpu.yml (hunk 8 above)
Comment (XinyueLunaFan) [reply .github/workflows/windows-build-xpu.yml]: Got it!

**Comment by XinyueLunaFan** [comment]
> Do we need to build against all default AOT targets for CI/CD?

Do we only need to build on MTL only and what is the default now?

File: .ci/pytorch/win-test-helpers/installation-helpers/install_oneapi_bundle.bat (hunk 6 above)
Comment (XinyueLunaFan) [reply .ci/pytorch/win-test-helpers/installation-helpers/install_oneapi_bundle.bat:2]: Whenever 0.5.3 publicly available, I will update it. Thank you.

**Comment by atalman** [comment]
hi @XinyueLunaFan VS2022 should already be available in the AMI we use for CI/CD. here is the installation step: https://github.com/pytorch/test-infra/blob/3d36a79aca555204002624459c7a2878a8711e39/aws/ami/windows/windows.pkr.hcl#L82

File: .ci/pytorch/win-test-helpers/build_pytorch.bat [hunk 11]
```diff
@@ -110,6 +118,25 @@ if "%USE_CUDA%"=="1" (
...
+  :: Install oneAPI bundle
+  if exist "C:\Program Files (x86)\Intel\oneAPI\setvars.bat" (
+      echo "oneAPI bundle already installed"
+  ) else (
+      echo "setvars.bat not found, installing oneAPI bundle"
+      call .ci\pytorch\win-test-helpers\installation-helpers\install_oneapi_bundle.bat
+      echo "oneAPI bundle installed"
+  )
+
+  :: Set up conda environment
```
Comment (chuanqi129) [inline .ci/pytorch/win-test-helpers/build_pytorch.bat:131]: Line 29 has `call %INSTALLER_DIR%\activate_miniconda3.bat`, why still need this part? Is there any special needs for xpu env setup?

File: .ci/pytorch/win-test-helpers/build_pytorch.bat [hunk 12]
```diff
@@ -110,6 +118,25 @@ if "%USE_CUDA%"=="1" (
   set CMAKE_CUDA_COMPILER_LAUNCHER=%TMP_DIR%/bin/randomtemp.exe;%TMP_DIR%\bin\sccache.exe
 )
 
+if "%USE_XPU%"=="1" (
+  :: Install oneAPI bundle
+  if exist "C:\Program Files (x86)\Intel\oneAPI\setvars.bat" (
+      echo "oneAPI bundle already installed"
+  ) else (
+      echo "setvars.bat not found, installing oneAPI bundle"
+      call .ci\pytorch\win-test-helpers\installation-helpers\install_oneapi_bundle.bat
```
Comment (chuanqi129) [inline .ci/pytorch/win-test-helpers/build_pytorch.bat:127]:
```suggestion
      call %INSTALLER_DIR%\install_xpu.bat
```

File: .ci/pytorch/win-test-helpers/build_pytorch.bat [hunk 13]
```diff
@@ -76,6 +76,14 @@ set PATH=%CUDA_PATH%\bin;%CUDA_PATH%\libnvvp;%PATH%
 
 :cuda_build_end
 
+if not "%USE_XPU%"=="1" goto xpu_build_end
+
+set PYTHON_VERSION='3.10'
+set VS2022INSTALLDIR='C:\Program Files (x86)\Microsoft Visual Studio\2022\BuildTools'
+set CONDA_ENV='xpu_venv'
```
Comment (chuanqi129) [inline .ci/pytorch/win-test-helpers/build_pytorch.bat:83]: Same here, why we need this var for XPU? I think for XPU, we just need to set MSVC 2022 only, others can we reuse existed env?

File: .ci/pytorch/win-test-helpers/installation-helpers/install_oneapi_bundle.bat [hunk 14]
```diff
@@ -0,0 +1,37 @@
...
+set INTEL_ONEAPI_PYTORCH_BUNDLE_DISPLAY_NAME=intel.oneapi.win.intel-for-pytorch-gpu-dev.product
+set INSTALL_FRESH_ONEAPI=1
+
+if not exist "%ONEAPI_PARENT_DIR%\oneAPI" (
+    set INSTALL_FRESH_ONEAPI=1
```
Comment (chuanqi129) [inline .ci/pytorch/win-test-helpers/installation-helpers/install_oneapi_bundle.bat:7]: =0? typo?

File: .ci/pytorch/win-test-helpers/installation-helpers/install_oneapi_bundle.bat [hunk 15]
```diff
@@ -0,0 +1,37 @@
...
+)
+
+
+if "%INSTALL_FRESH_ONEAPI%"=="1" (
+    IF EXIST "%ONEAPI_PARENT_DIR%\oneAPI\Installer\installer.exe" (
```
Comment (chuanqi129) [inline .ci/pytorch/win-test-helpers/installation-helpers/install_oneapi_bundle.bat:12]: Will we compare the bundle version between plan to install and installed, when the versions are same, we will not remove and reinstall? If not, let's simply the script logic, if there is installed bundle, always remove it and install the specific bundle in this script.

File: .ci/pytorch/win-test-helpers/build_pytorch.bat (hunk 12 above)
Comment (chuanqi129) [reply .ci/pytorch/win-test-helpers/build_pytorch.bat:127]: And move this part before line 33

File: .ci/pytorch/win-test-helpers/installation-helpers/install_oneapi_bundle.bat
Comment (chuanqi129) [inline .ci/pytorch/win-test-helpers/installation-helpers/install_oneapi_bundle.bat:1]: Let's rename it as `install_xpu.bat`. And please refer the steps in https://www.intel.com/content/www/us/en/developer/articles/tool/pytorch-prerequisites-for-intel-gpu/2-5.html, and add this link in this script

File: .github/workflows/xpu.yml [hunk 16]
```diff
@@ -37,3 +37,10 @@ jobs:
       build-environment: linux-jammy-xpu-py3.8
       docker-image: ${{ needs.linux-jammy-xpu-py3_8-build.outputs.docker-image }}
       test-matrix: ${{ needs.linux-jammy-xpu-py3_8-build.outputs.test-matrix }}
+
+  windows-xpu-build:
+    name: windows-xpu-py3.10
```
Comment (chuanqi129) [inline .github/workflows/xpu.yml:42]:
```suggestion
    name: win-vs2022-xpu-py3
```

File: .github/workflows/xpu.yml [hunk 17]
```diff
@@ -37,3 +37,10 @@ jobs:
...
+    uses: ./.github/workflows/_win-build.yml
+    with:
+      build-environment: xpu-py3.10
```
Comment (chuanqi129) [inline .github/workflows/xpu.yml:45]:
```suggestion
      build-environment: win-vs2022-xpu-py3
```

File: .github/workflows/_win-build.yml [hunk 18]
```diff
@@ -11,6 +11,16 @@ on:
         required: true
         type: string
         description: What CUDA version to build with, "cpu" for none.
+      use-xpu:
+        required: false
+        type: boolean
+        default: false
+        description: If set, build with XPU support.
+      vc-year:
+        required: true
```
Comment (chuanqi129) [inline .github/workflows/_win-build.yml:20]:
```suggestion
        required: false
```

File: .ci/pytorch/win-test-helpers/installation-helpers/install_xpu.bat [hunk 19]
```diff
@@ -0,0 +1,55 @@
+REM Description: Install Intel oneAPI Pytorch Bundle on Windows
```
Comment (chuanqi129) [inline .ci/pytorch/win-test-helpers/installation-helpers/install_xpu.bat:1]:
```suggestion
REM Description: Install Intel Support Packages on Windows
```

Comment (chuanqi129) [reply .ci/pytorch/win-test-helpers/installation-helpers/install_xpu.bat:1]: Add todo in here to add GPU driver installation function

**Review by chuanqi129 — APPROVED** [review]
LGTM. @XinyueLunaFan, please update the description of this PR accordingly. Also works for https://github.com/pytorch/pytorch/issues/114850

**Comment by chuanqi129** [comment]
Hi @atalman, could you please help to review this xpu windows build ci enabling PR? Thanks

**Comment by chuanqi129** [comment]
@XinyueLunaFan Let's update the intel support package version to 0.5.3

**Comment by atalman** [comment]
Lets wait for binaries build to complete

File: .ci/pytorch/win-test-helpers/build_pytorch.bat [hunk 20]
```diff
@@ -24,6 +24,11 @@ call %INSTALLER_DIR%\install_sccache.bat
 if errorlevel 1 goto fail
 if not errorlevel 0 goto fail
 
+if "%USE_XPU%"=="1" (
+  :: Install oneAPI bundle
+  call %INSTALLER_DIR%\install_xpu.bat
+)
+
```
Comment (chuanqi129) [inline .ci/pytorch/win-test-helpers/build_pytorch.bat:31]: Unfortunately, the current change can't build xpu wheel actually, refer https://github.com/pytorch/pytorch/actions/runs/10497376953/job/29080090269?pr=133151#step:11:623. Please follow the official bkc to activate bundle env

**Review by chuanqi129 — CHANGES_REQUESTED** [review]

File: .ci/pytorch/win-test-helpers/build_pytorch.bat (hunk 20 above)
Comment (XinyueLunaFan) [reply .ci/pytorch/win-test-helpers/build_pytorch.bat:31]: Thanks for pointing out. I have added env var set up in the script as well.

File: .ci/pytorch/win-test-helpers/installation-helpers/install_xpu.bat [hunk 21]
```diff
@@ -0,0 +1,109 @@
...
+    )
+
+    curl -o xpu_pti.exe --retry 3 --retry-all-errors -k %XPU_PTI_URL%
+    start /wait "Intel PTI Installer" "xpu_pti.exe" --action=install --eula=accept --silent --log-dir install_pti
+    if errorlevel 1 exit /b
+    if not errorlevel 0 exit /b
+    del xpu_pti.exe
+)
+
+REM Set environment variables
```
Comment (chuanqi129) [inline .ci/pytorch/win-test-helpers/installation-helpers/install_xpu.bat:107]: Move out the bundle env activate part into `build_pytorch.bat`

File: .ci/pytorch/win-test-helpers/installation-helpers/install_xpu.bat [hunk 22]
```diff
@@ -0,0 +1,109 @@
...
+        if not errorlevel 0 exit /b
+    )
+
+    curl -o xpu_pti.exe --retry 3 --retry-all-errors -k %XPU_PTI_URL%
+    start /wait "Intel PTI Installer" "xpu_pti.exe" --action=install --eula=accept --silent --log-dir install_pti
+    if errorlevel 1 exit /b
+    if not errorlevel 0 exit /b
+    del xpu_pti.exe
+)
+
```
Comment (chuanqi129) [inline .ci/pytorch/win-test-helpers/installation-helpers/install_xpu.bat:106]: Let's align the xpu install with CD part https://github.com/pytorch/builder/pull/1975. Please note,

- [ ] Driver can't be installed on no XPU node. Build also don't need install driver

- [ ] Enhance the driver installation part, for now the installation seems cause hang, I didn't test for it.
```suggestion
@echo on
REM Description: Install Intel Support Packages on Windows
REM BKM reference: https://www.intel.com/content/www/us/en/developer/articles/tool/pytorch-prerequisites-for-intel-gpu/2-5.html

set XPU_INSTALL_MODE=%~1
if "%XPU_INSTALL_MODE%"=="" goto xpu_bundle_install_start
if "%XPU_INSTALL_MODE%"=="bundle" goto xpu_bundle_install_start
if "%XPU_INSTALL_MODE%"=="driver" goto xpu_driver_install_start
if "%XPU_INSTALL_MODE%"=="all" goto xpu_driver_install_start

:arg_error

echo Illegal XPU installation mode. The value can be "bundle"/"driver"/"all"
echo If keep the value as space, will use default "bundle" mode
exit /b 1

:xpu_driver_install_start
:: TODO Need more testing for driver installation
set XPU_DRIVER_LINK=https://downloadmirror.intel.com/830975/gfx_win_101.5972.exe
curl -o xpu_driver.exe --retry 3 --retry-all-errors -k %XPU_DRIVER_LINK%
echo "XPU Driver installing..."
start /wait "Intel XPU Driver Installer" "xpu_driver.exe"
if errorlevel 1 exit /b 1
del xpu_driver.exe
if "%XPU_INSTALL_MODE%"=="driver" goto xpu_install_end

:xpu_bundle_install_start

set XPU_BUNDLE_PARENT_DIR=C:\Program Files (x86)\Intel\oneAPI
set XPU_BUNDLE_URL=https://registrationcenter-download.intel.com/akdlm/IRC_NAS/9d1a91e2-e8b8-40a5-8c7f-5db768a6a60c/w_intel-for-pytorch-gpu-dev_p_0.5.3.37_offline.exe
set XPU_PTI_URL=https://registrationcenter-download.intel.com/akdlm/IRC_NAS/9d1a91e2-e8b8-40a5-8c7f-5db768a6a60c/w_intel-pti-dev_p_0.9.0.37_offline.exe
set XPU_BUNDLE_VERSION=0.5.3+31
set XPU_PTI_VERSION=0.9.0+36
set XPU_BUNDLE_PRODUCT_NAME=intel.oneapi.win.intel-for-pytorch-gpu-dev.product
set XPU_PTI_PRODUCT_NAME=intel.oneapi.win.intel-pti-dev.product
set XPU_BUNDLE_INSTALLED=0
set XPU_PTI_INSTALLED=0
set XPU_BUNDLE_UNINSTALL=0
set XPU_PTI_UNINSTALL=0

:: Check if XPU bundle is target version or already installed
if exist "%XPU_BUNDLE_PARENT_DIR%\Installer\installer.exe" goto xpu_bundle_ver_check
goto xpu_bundle_install

:xpu_bundle_ver_check

"%XPU_BUNDLE_PARENT_DIR%\Installer\installer.exe" --list-products > xpu_bundle_installed_ver.log

for /f "tokens=1,2" %%a in (xpu_bundle_installed_ver.log) do (
    if "%%a"=="%XPU_BUNDLE_PRODUCT_NAME%" (
        echo %%a Installed Version: %%b
	    set XPU_BUNDLE_INSTALLED=1
        if not "%XPU_BUNDLE_VERSION%"=="%%b" (
            start /wait "Installer Title" "%XPU_BUNDLE_PARENT_DIR%\Installer\installer.exe" --action=remove --eula=accept --silent --product-id %XPU_BUNDLE_PRODUCT_NAME% --product-ver %%b --log-dir uninstall_bundle
            set XPU_BUNDLE_UNINSTALL=1
        )
    )
    if "%%a"=="%XPU_PTI_PRODUCT_NAME%" (
        echo %%a Installed Version: %%b
	    set XPU_PTI_INSTALLED=1
        if not "%XPU_PTI_VERSION%"=="%%b" (
            start /wait "Installer Title" "%XPU_BUNDLE_PARENT_DIR%\Installer\installer.exe" --action=remove --eula=accept --silent --product-id %XPU_PTI_PRODUCT_NAME% --product-ver %%b --log-dir uninstall_bundle
            set XPU_PTI_UNINSTALL=1
        )
    )
)
if errorlevel 1 exit /b 1
if exist xpu_bundle_installed_ver.log del xpu_bundle_installed_ver.log
if "%XPU_BUNDLE_INSTALLED%"=="0" goto xpu_bundle_install
if "%XPU_BUNDLE_UNINSTALL%"=="1" goto xpu_bundle_install
if "%XPU_PTI_INSTALLED%"=="0" goto xpu_pti_install
if "%XPU_PTI_UNINSTALL%"=="1" goto xpu_pti_install
goto xpu_install_end

:xpu_bundle_install

curl -o xpu_bundle.exe --retry 3 --retry-all-errors -k %XPU_BUNDLE_URL%
echo "XPU Bundle installing..."
start /wait "Intel Pytorch Bundle Installer" "xpu_bundle.exe" --action=install --eula=accept --silent --log-dir install_bundle
if errorlevel 1 exit /b 1
del xpu_bundle.exe

:xpu_pti_install

curl -o xpu_pti.exe --retry 3 --retry-all-errors -k %XPU_PTI_URL%
echo "XPU PTI installing..."
start /wait "Intel PTI Installer" "xpu_pti.exe" --action=install --eula=accept --silent --log-dir install_bundle
if errorlevel 1 exit /b 1
del xpu_pti.exe

:xpu_install_end
```

**Review by chuanqi129 — APPROVED** [review]

**Review by atalman — APPROVED** [review]
lgtm

**Pull request closed without merging**